### PR TITLE
Specify @throws for `node.path` methods

### PR DESCRIPTION
**Pull request by mbelsky**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- ~[Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.~
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [url here](https://nodejs.org/docs/latest-v16.x/api/path.html)
- ~If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~

### CODE REVIEW

File: types/node/path.d.ts [hunk 1]
```diff
@@ -135,15 +145,16 @@ declare module 'path' {
             /**
              * Returns an object from a path string - the opposite of format().
              *
-             * @param pathString path to evaluate.
+             * @param p path to evaluate.
+             * @throws {TypeError} if path is not a string.
              */
             parse(p: string): ParsedPath;
```
Comment (peterblazejewicz) [inline types/node/path.d.ts:151]:
```suggestion
             * @param path path to evaluate.
             * @throws {TypeError} if `path` is not a string.
             */
            parse(p: string): ParsedPath;
```

File: types/node/v14/path.d.ts [hunk 2]
```diff
@@ -74,24 +75,31 @@ declare module 'path' {
...
              */
-            resolve(...pathSegments: string[]): string;
+            resolve(...paths: string[]): string;
             /**
              * Determines whether {path} is an absolute path. An absolute path will always resolve to the same location, regardless of the working directory.
              *
+             * If the given {path} is a zero-length string, `false` will be returned.
+             *
              * @param path path to test.
+             * @throws {TypeError} if path is not a string.
```
Comment (peterblazejewicz) [inline types/node/v14/path.d.ts:88]:
```suggestion
             * @throws {TypeError} if `path` is not a string.
```

File: types/node/v14/path.d.ts [hunk 3]
```diff
@@ -74,24 +75,31 @@ declare module 'path' {
...
+             * @throws {TypeError} if either {from} or {to} is not a string.
              */
             relative(from: string, to: string): string;
             /**
              * Return the directory name of a path. Similar to the Unix dirname command.
              *
              * @param p the path to evaluate.
+             * @throws {TypeError} if path is not a string.
              */
             dirname(p: string): string;
```
Comment (peterblazejewicz) [inline types/node/v14/path.d.ts:104]:
```suggestion
             * @param path the path to evaluate.
             * @throws {TypeError} if `path `is not a string.
             */
            dirname(path: string): string;
```

File: types/node/v14/path.d.ts [hunk 4]
```diff
@@ -100,13 +108,15 @@ declare module 'path' {
...
             basename(p: string, ext?: string): string;
             /**
              * Return the extension of the path, from the last '.' to end of string in the last portion of the path.
-             * If there is no '.' in the last portion of the path or the first character of it is '.', then it returns an empty string
+             * If there is no '.' in the last portion of the path or the first character of it is '.', then it returns an empty string.
              *
              * @param p the path to evaluate.
+             * @throws {TypeError} if path is not a string.
              */
             extname(p: string): string;
```
Comment (peterblazejewicz) [inline types/node/v14/path.d.ts:121]:
```suggestion
             * @param path the path to evaluate.
             * @throws {TypeError} if `path` is not a string.
             */
            extname(path: string): string;
```

File: types/node/v16/path.d.ts [hunk 5]
```diff
@@ -89,24 +90,31 @@ declare module 'path' {
...
+             * @throws {TypeError} if any of the arguments is not a string.
              */
             resolve(...pathSegments: string[]): string;
             /**
              * Determines whether {path} is an absolute path. An absolute path will always resolve to the same location, regardless of the working directory.
              *
+             * If the given {path} is a zero-length string, `false` will be returned.
+             *
              * @param path path to test.
+             * @throws {TypeError} if path is not a string.
```
Comment (peterblazejewicz) [inline types/node/v16/path.d.ts:103]:
```suggestion
             * @throws {TypeError} if `path` is not a string.
```

File: types/node/v16/path.d.ts [hunk 6]
```diff
@@ -89,24 +90,31 @@ declare module 'path' {
...
              * At times we have two absolute paths, and we need to derive the relative path from one to the other. This is actually the reverse transform of path.resolve.
+             *
+             * @throws {TypeError} if either {from} or {to} is not a string.
              */
             relative(from: string, to: string): string;
             /**
              * Return the directory name of a path. Similar to the Unix dirname command.
              *
              * @param p the path to evaluate.
+             * @throws {TypeError} if path is not a string.
```
Comment (peterblazejewicz) [inline types/node/v16/path.d.ts:117]:
```suggestion
             * @throws {TypeError} if `path` is not a string.
```

File: types/node/v16/path.d.ts [hunk 7]
```diff
@@ -115,13 +123,15 @@ declare module 'path' {
              *
              * @param p the path to evaluate.
              * @param ext optionally, an extension to remove from the result.
+             * @throws {TypeError} if path is not a string or if ext is given and is not a string.
```
Comment (peterblazejewicz) [inline types/node/v16/path.d.ts:126]:
```suggestion
             * @throws {TypeError} if `path` is not a string or if `ext` is given and is not a string.
```

File: types/node/v16/path.d.ts [hunk 8]
```diff
@@ -115,13 +123,15 @@ declare module 'path' {
              *
              * @param p the path to evaluate.
```
Comment (peterblazejewicz) [inline types/node/v16/path.d.ts:124]:
```suggestion
             * @param path the path to evaluate.
```

File: types/node/v16/path.d.ts [hunk 9]
```diff
@@ -115,13 +123,15 @@ declare module 'path' {
...
              * @param ext optionally, an extension to remove from the result.
+             * @throws {TypeError} if path is not a string or if ext is given and is not a string.
              */
             basename(p: string, ext?: string): string;
```
Comment (peterblazejewicz) [inline types/node/v16/path.d.ts:128]:
```suggestion
            basename(path: string, ext?: string): string;
```

File: types/node/v16/path.d.ts [hunk 10]
```diff
@@ -135,15 +145,16 @@ declare module 'path' {
             /**
              * Returns an object from a path string - the opposite of format().
              *
-             * @param pathString path to evaluate.
+             * @param p path to evaluate.
+             * @throws {TypeError} if path is not a string.
              */
             parse(p: string): ParsedPath;
```
Comment (peterblazejewicz) [inline types/node/v16/path.d.ts:151]:
```suggestion
             * @param path path to evaluate.
             * @throws {TypeError} if `path` is not a string.
             */
            parse(path: string): ParsedPath;
```

File: types/node/v16/path.d.ts [hunk 11]
```diff
@@ -115,13 +123,15 @@ declare module 'path' {
...
             basename(p: string, ext?: string): string;
             /**
              * Return the extension of the path, from the last '.' to end of string in the last portion of the path.
-             * If there is no '.' in the last portion of the path or the first character of it is '.', then it returns an empty string
+             * If there is no '.' in the last portion of the path or the first character of it is '.', then it returns an empty string.
              *
              * @param p the path to evaluate.
+             * @throws {TypeError} if path is not a string.
              */
             extname(p: string): string;
```
Comment (peterblazejewicz) [inline types/node/v16/path.d.ts:136]:
```suggestion
             * @param path the path to evaluate.
             * @throws {TypeError} if `path` is not a string.
             */
            extname(path: string): string;
```

File: types/node/path.d.ts [hunk 12]
```diff
@@ -115,13 +123,15 @@ declare module 'path' {
              *
              * @param p the path to evaluate.
              * @param ext optionally, an extension to remove from the result.
+             * @throws {TypeError} if path is not a string or if ext is given and is not a string.
              */
             basename(p: string, ext?: string): string;
```
Comment (peterblazejewicz) [inline types/node/path.d.ts:128]:
```suggestion
             * @param path the path to evaluate.
             * @param ext optionally, an extension to remove from the result.
             * @throws {TypeError} if `path` is not a string or if `ext` is given and is not a string.
             */
            basename(path: string, ext?: string): string;
```

File: types/node/path.d.ts [hunk 13]
```diff
@@ -89,24 +90,31 @@ declare module 'path' {
...
+             * @throws {TypeError} if either {from} or {to} is not a string.
              */
             relative(from: string, to: string): string;
             /**
              * Return the directory name of a path. Similar to the Unix dirname command.
              *
              * @param p the path to evaluate.
+             * @throws {TypeError} if path is not a string.
              */
             dirname(p: string): string;
```
Comment (peterblazejewicz) [inline types/node/path.d.ts:119]:
```suggestion
             * @param path the path to evaluate.
             * @throws {TypeError} if path is not a string.
             */
            dirname(path: string): string;
```

**Comment by peterblazejewicz** [comment]
added some inline comments to make those consistent

**Comment by mbelsky** [comment]
@peterblazejewicz thank you for finding them! Fixed

File: types/node/path.d.ts [hunk 14]
```diff
@@ -69,18 +69,19 @@ declare module 'path' {
              * Normalize a string path, reducing '..' and '.' parts.
              * When multiple slashes are found, they're replaced by a single one; when the path contains a trailing slash, it is preserved. On Windows backslashes are used.
              *
-             * @param p string path to normalize.
+             * @param path string path to normalize.
+             * @throws {TypeError} if `path` is not a string.
```
Comment (SimonSchick) [inline types/node/path.d.ts:73]: On the fence on this, TS itself (sadly) doesn't have checked exception and TS already checks that all segments are strings, this feels redundant in a type checked context.

**Review by SimonSchick — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

**Comment by mbelsky** [comment]
Ready to merge

**Pull request merged**